Documentation/design-docs/framework-version-resolution.md

```diff
@@ -0,0 +1,299 @@
+* `LatestMinor` -- Roll forward to highest minor version, even if requested minor version is present.
+* `LatestMajor` -- Roll forward to highest major and highest minor version, even if requested major is present.
+* `Disable` -- Do not roll forward. Only bind to specified version. This policy is not recommended for general use since it disables the ability to roll forward to the latest patches. It is only recommended for testing.
+
+Roll-forward setting can be specified in these places:
+1. `.runtimeconfig.json` - `runtimeOptions` property `rollForward` - at this level the setting applies to all framework references in that `.runtimeconfig.json`. For example:
+``` json
+{
+  "runtimeOptions": {
+    "tfm": "netcoreapp3.0",
+    "rollForward": "major",
+    "frameworks": [
+      {
+        "name": "Microsoft.NETCore.App",
+        "version": "3.0.1"
+      }
+    ]
+  }
+}
+```
+
+2. `.runtimeconfig.json` - `framework` property `rollForward` - each framework reference can specify its own setting. For example:
+``` json
+{
+  "runtimeOptions": {
+    "tfm": "netcoreapp3.0",
+    "frameworks": [
+      {
+        "name": "Microsoft.NETCore.App",
+        "version": "3.0.1",
+        "rollForward": "major"
+      }
+    ]
+  }
+}
+```
+
+3. Environment variable `DOTNET_ROLL_FORWARD` - For example:
+```
+set DOTNET_ROLL_FORWARD=LatestMajor
+```
+
+4. Command line argument `--roll-forward` - For example:
+```
+dotnet --roll-forward LatestMinor app.dll
+```
+
+For the framework references specified by the application itself:  
+The order above defines precedence. Later scopes take precedence over earlier ones. So command line wins over everything, environment variable wins over `.runtimeconfig.json` and per-framework setting wins over config-wide setting.
+
+For dependent frameworks:
+Each dependent framework (for example FX1) is loaded with an effective roll-forward value (the one used to resolve that framework). If FX1 has additional framework references those will be loaded using this same roll-forward value as the default, that is unless FX1 specifies different value in its `.runtimeconfig.json` or there's a higher precedence override (environment variable `DOTNET_ROLL_FORWARD` or command line argument `--roll-forward`).
+
+
+## Pre-release versions
+Pre-release version is a version which has a pre-release part, for example `3.0.0-preview4-27415-15`. Everything after the `-` (dash) character is a pre-release identifier. Per [semantic versioning rules](https://semver.org/) pre-release versions are ordered before the same version without any pre-release part. So `3.0.0-preview4-27415-15` comes before `3.0.0`.
+
+TODO: Figure out the equality of pre-release and normal versions. That is, per SemVer rules, is it acceptable to resolve reference to `3.0.0` with a pre-release `3.0.0-preview`?
```

Comment:
SemVer 2.0.0-rc2 had this language 
```Pre-release versions satisfy but have a lower precedence than the associated normal version```.  

In the SemVer2.0.0 release that language was replaced with 
```A pre-release version indicates that the version is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version```

So apparently it was controversial.  
I would recommend
+ `DOTNET_ROLL_FORWARD_TO_PRERELEASE=0` **DO NOT** resolve reference to `3.0.0` with a pre-release `3.0.0-preview`
+ `DOTNET_ROLL_FORWARD_TO_PRERELEASE=1` resolve reference to `3.0.0` with a pre-release `3.0.0-preview`
proposed below.  When is is disabled, it doesn't treat prerelease as sati

Reply:
That sounds good!

Reply:
I'd prefer if the flag did not cause 3.0.0 ref to resolve to 3.0.0-preview. It's not in the use case I had for this flag of testing previews. I don't think NuGet interprests versions and I find it surprising. This can be a massive downgrade in practice for our own bits as tons of functionality is added between 3.0-preview1,2,3,4,5,6,...N -> RTM.

Reply:
@nguerrera If an app specifies `3.0.0` framework, but only `3.0.0-preview` is installed.  You want it to fail?  How do we test that scenario before release?

Reply:
We don't generally have any apps that specify 3.0.0 before release. The reason is that you need a 3.0.0 SDK to produce that from TFM netcoreapp3.0 (or take matters into your hands in a variety of ways). The big pain point that I raised starts at x.y.1.  Then there are tons of apps that specify x.y.0 and we want to test that x.y.n-preview doesn't break them.

The scenarios I want to test do not involve downgrades. I don't want to test that an app that says it needs 3.0 RTM can run on builds before RTM. I want to test that it can run on builds produced after RTM. 

That said, I could see the logical consistency of applying the general principle of "the feature is for testing that x.y.z-preview behaves as x.y.z" here. I no longer oppose this. :)

Reply:
If we don't need it, I am not opposed to leaving it out either.  I don't think it will naturally fall out of the implementation.

Reply:
I guess to implement this we would need search for `3.0.0-0` (if I am thinking correctly is the youngest SemVer2 3.0.0 prerelease) instead of `3.0.0` when the flag was enabled.  We would still roll forward from there....

Reply:
Based on the discussion here and the current product behavior I've change the document to follow SemVer2 exactly - which means that `3.0.0-preview` does NOT satisfy a `3.0.0` reference (it would be a downgrade per SemVer2 rules).

I think it also matches expectation in the sense that `3.0.0-preview` may not implement all the features of `3.0.0`.